File: PSVRPlayer/Handy/HandyClient.cs
using System;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace PSVRPlayer.Handy;

/// <summary>
/// The Handy API v2 client.
/// HSSP (server-sync) mode: upload funscript → sync time → play/stop with timestamp.
/// </summary>
public sealed class HandyClient : IDisposable
{
    private const string Base   = "https://www.handyfeeling.com/api/handy/v2";
    private const string Upload = "https://www.handyfeeling.com/api/script/v0/upload";

    private readonly HttpClient _http = new();
    private string _key = string.Empty;

    // HSSP mode value = 4 (PSVRFramework / Handy API v2)
    private const int ModeHSSP = 4;

    public string? ScriptUrl     { get; private set; }
    public double  TimeOffsetMs  { get; private set; }  // local+offset = server time
    public bool    IsReady       { get; private set; }

    public void SetKey(string key) => _key = key.Trim();

    private HttpRequestMessage Req(HttpMethod m, string path)
    {
        var r = new HttpRequestMessage(m, Base + path);
        r.Headers.Add("X-Connection-Key", _key);
        r.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        return r;
    }

    // ── Connection ────────────────────────────────────────────────────────

    public async Task<bool> CheckConnectionAsync()
    {
        using var res = await _http.SendAsync(Req(HttpMethod.Get, "/connected"));
        res.EnsureSuccessStatusCode();
        using var doc = await JsonDocument.ParseAsync(await res.Content.ReadAsStreamAsync());
        return doc.RootElement.GetProperty("connected").GetBoolean();
    }

    // ── Time sync (Handy server-clock alignment) ──────────────────────────

    public async Task SyncTimeAsync(int iterations = 30, Action<int, int>? progress = null)
    {
        var offsets = new double[iterations];
        int valid   = 0;

        for (int i = 0; i < iterations; i++)
        {
            long t0 = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            try
            {
                using var res = await _http.SendAsync(Req(HttpMethod.Get, "/servertime"));
                long t1 = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
                using var doc = await JsonDocument.ParseAsync(await res.Content.ReadAsStreamAsync());
                long sv = doc.RootElement.GetProperty("serverTime").GetInt64();
                offsets[valid++] = sv - t1 + (t1 - t0) / 2.0;
            }
            catch { /* skip */ }
            progress?.Invoke(i + 1, iterations);
        }

        if (valid < 5) throw new Exception("タイムシンク失敗 (5回未満成功)");

        Array.Sort(offsets, 0, valid);
        int trim = Math.Max(1, valid / 10);
        double sum = 0;
        for (int i = trim; i < valid - trim; i++) sum += offsets[i];
        TimeOffsetMs = sum / (valid - 2 * trim);
    }

    private long ServerNow() => DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() + (long)TimeOffsetMs;

    // ── Script upload ─────────────────────────────────────────────────────

    public async Task<string> UploadScriptAsync(FunscriptPlayer funscript)
    {
        var json = JsonSerializer.Serialize(new { version = "1.0", actions = funscript.Actions });
        using var content = new MultipartFormDataContent();
        content.Add(new ByteArrayContent(Encoding.UTF8.GetBytes(json))
        {
            Headers = { ContentType = new MediaTypeHeaderValue("application/json") }
        }, "file", "script.funscript");

        using var res = await _http.PostAsync(Upload, content);
        res.EnsureSuccessStatusCode();
        using var doc = await JsonDocument.ParseAsync(await res.Content.ReadAsStreamAsync());
        ScriptUrl = doc.RootElement.GetProperty("url").GetString()
            ?? throw new Exception("アップロードレスポンスにURLがありません");
        return ScriptUrl;
    }

    // ── HSSP setup ────────────────────────────────────────────────────────

    public async Task SetupHSSPAsync(string? url = null)
    {
        url ??= ScriptUrl ?? throw new InvalidOperationException("先にアップロードしてください");

        // Set mode to HSSP
        var modeReq = Req(HttpMethod.Put, "/mode");
        modeReq.Content = new StringContent(JsonSerializer.Serialize(new { mode = ModeHSSP }),
            Encoding.UTF8, "application/json");
        (await _http.SendAsync(modeReq)).EnsureSuccessStatusCode();

        // Configure script URL + timeout
        var setupReq = Req(HttpMethod.Put, "/hssp/setup");
        setupReq.Content = new StringContent(JsonSerializer.Serialize(new { url, timeout = 10000 }),
            Encoding.UTF8, "application/json");
        (await _http.SendAsync(setupReq)).EnsureSuccessStatusCode();

        IsReady = true;
    }

    // ── Playback control ──────────────────────────────────────────────────

    public async Task PlayAsync(double videoTimeSec)
    {
        if (!IsReady) return;
        const int estimatedDelay = 150; // ms latency between API call and device acting
        long startAt = ServerNow() + estimatedDelay - (long)(videoTimeSec * 1000);

        var req = Req(HttpMethod.Put, "/hssp/play");
        req.Content = new StringContent(
            JsonSerializer.Serialize(new { serverTime = startAt, playbackRate = 1.0 }),
            Encoding.UTF8, "application/json");
        await _http.SendAsync(req);
    }

    public async Task StopAsync()
    {
        if (!IsReady) return;
        await _http.SendAsync(Req(HttpMethod.Put, "/hssp/stop"));
    }

    public async Task SeekAsync(double videoTimeSec, bool isPlaying)
    {
        await StopAsync();
        if (isPlaying) await PlayAsync(videoTimeSec);
    }

    public void Dispose() => _http.Dispose();
}

File: src/HandyClient.js
/**
 * The Handy API v2 client.
 * API docs: https://www.handyfeeling.com/api/handy/v2/docs
 *
 * Authentication: X-Connection-Key header
 * HSSP mode (mode=4): upload script → setup → play/stop with server-time sync
 */

const BASE_URL = 'https://www.handyfeeling.com/api/handy/v2';
const SCRIPT_UPLOAD_URL = 'https://www.handyfeeling.com/api/script/v0/upload';

// Mode enum (Handy API v2)
const Mode = {
  HAMP: 0,
  HSSP: 4,
};

export class HandyClient {
  constructor() {
    this.connectionKey = '';
    this.serverTimeOffset = 0; // local + offset = server time
    this.scriptUrl = null;
    this.isConnected = false;
    this.isReady = false; // true after sync + script setup
    this._onStatusChange = null;
  }

  setConnectionKey(key) {
    this.connectionKey = key.trim();
  }

  onStatusChange(cb) {
    this._onStatusChange = cb;
  }

  _emit(status) {
    this._onStatusChange?.(status);
  }

  get _headers() {
    return {
      'X-Connection-Key': this.connectionKey,
      'Accept': 'application/json',
      'Content-Type': 'application/json',
    };
  }

  async checkConnection() {
    if (!this.connectionKey) throw new Error('接続キーが入力されていません');

    const res = await fetch(`${BASE_URL}/connected`, { headers: this._headers });
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    const data = await res.json();
    this.isConnected = data.connected === true;
    return this.isConnected;
  }

  /**
   * Sync local clock to Handy server clock.
   * Sends N requests, trims outliers, averages the offset.
   */
  async syncTime(iterations = 30, onProgress = null) {
    const offsets = [];

    for (let i = 0; i < iterations; i++) {
      const t0 = Date.now();
      try {
        const res = await fetch(`${BASE_URL}/servertime`, {
          headers: { 'X-Connection-Key': this.connectionKey, Accept: 'application/json' },
        });
        const t1 = Date.now();
        const { serverTime } = await res.json();
        const rtt = t1 - t0;
        // Estimated server time at receipt = serverTime (already the server's current time)
        // Local clock at receipt = t1
        // Offset = serverTime - t1 + rtt/2  (account for one-way latency)
        offsets.push(serverTime - t1 + rtt / 2);
      } catch (_) { /* skip failed requests */ }

      onProgress?.(i + 1, iterations);
    }

    if (offsets.length < 5) throw new Error('タイムシンクに失敗しました');

    offsets.sort((a, b) => a - b);
    const trim = Math.max(1, Math.floor(offsets.length * 0.1));
    const trimmed = offsets.slice(trim, offsets.length - trim);
    this.serverTimeOffset = trimmed.reduce((s, v) => s + v, 0) / trimmed.length;
    return this.serverTimeOffset;
  }

  serverNow() {
    return Date.now() + this.serverTimeOffset;
  }

  async _setMode(mode) {
    const res = await fetch(`${BASE_URL}/mode`, {
      method: 'PUT',
      headers: this._headers,
      body: JSON.stringify({ mode }),
    });
    if (!res.ok) throw new Error(`モード設定失敗 HTTP ${res.status}`);
  }

  /**
   * Upload funscript JSON to Handy's CDN. Returns hosted URL.
   */
  async uploadScript(funscriptJson) {
    const json = typeof funscriptJson === 'string' ? funscriptJson : JSON.stringify(funscriptJson);
    const blob = new Blob([json], { type: 'application/json' });
    const form = new FormData();
    form.append('file', blob, 'script.funscript');

    const res = await fetch(SCRIPT_UPLOAD_URL, { method: 'POST', body: form });
    if (!res.ok) throw new Error(`アップロード失敗 HTTP ${res.status}`);
    const data = await res.json();
    if (!data.url) throw new Error('アップロードレスポンスにURLがありません');
    this.scriptUrl = data.url;
    return data.url;
  }

  /**
   * Configure HSSP with the uploaded script URL.
   */
  async setupHSSP(url = this.scriptUrl) {
    if (!url) throw new Error('スクリプトURLがありません。先にアップロードしてください。');

    await this._setMode(Mode.HSSP);

    const res = await fetch(`${BASE_URL}/hssp/setup`, {
      method: 'PUT',
      headers: this._headers,
      body: JSON.stringify({ url, timeout: 10000 }),
    });
    if (!res.ok) throw new Error(`HSSP設定失敗 HTTP ${res.status}`);
    this.isReady = true;
  }

  /**
   * Start HSSP playback. videoTimeSec is the current video playback position in seconds.
   * The Handy will sync its script to match the video's current time.
   */
  async play(videoTimeSec) {
    if (!this.isReady) return;
    // estimatedDelay: time between sending API request and device acting on it (~150ms)
    const estimatedDelay = 150;
    const serverStart = this.serverNow() + estimatedDelay - videoTimeSec * 1000;

    const res = await fetch(`${BASE_URL}/hssp/play`, {
      method: 'PUT',
      headers: this._headers,
      body: JSON.stringify({ serverTime: Math.round(serverStart), playbackRate: 1.0 }),
    });
    if (!res.ok) throw new Error(`再生失敗 HTTP ${res.status}`);
  }

  async stop() {
    if (!this.isReady) return;
    const res = await fetch(`${BASE_URL}/hssp/stop`, {
      method: 'PUT',
      headers: this._headers,
    });
    if (!res.ok) throw new Error(`停止失敗 HTTP ${res.status}`);
  }

  async seek(videoTimeSec, isPlaying) {
    if (!this.isReady) return;
    await this.stop();
    if (isPlaying) await this.play(videoTimeSec);
  }
}
